src/core/net/ip6.cpp

```diff
@@ -977,6 +978,13 @@ Error Ip6::PassToHost(Message           &aMessage,
     // Do not pass IPv6 packets that exceed kMinimalMtu.
     VerifyOrExit(aMessage.GetLength() <= kMinimalMtu, error = kErrorDrop);
 
+    // When the packet is classified as 'receive' or it has a multicast destination we should always forward it to host.
+    // When the packet's source is mesh-local and the destination is not on the host, we should not forward it to host.
+    VerifyOrExit(aReceive || aMessageInfo.GetSockAddr().IsMulticast() ||
+                     !Get<Mle::Mle>().IsMeshLocalAddress(aMessageInfo.GetPeerAddr()) ||
+                     Get<ThreadNetif>().GetUnicastAddresses().ContainsMatching(aMessageInfo.GetSockAddr()),
+                 error = kErrorDrop);
```

Comment:
Do we need the check `aMessageInfo.GetSockAddr().IsMulticast()` (for when the destination is multicast address)?

This whole check applies when source is mesh-local, then:
- If this destination is for a multicast address that Thread interface is subscribed to, then `aReceive` would be set already (which causes us to always pass to host).
- If thread interface itself is not subscribed to this multicast address, then `aReceive` would not be set and in this case, I think we don't want to pass it to host (i.e. source is mesh-local but dest is multicat address not on device?)